operator/pkg/ciliumidentity/reconciler.go

```
@@ -0,0 +1,444 @@
+		podStore:        podStore,
+		cidStore:        cidStore,
+		cepStore:        cepStore,
+		cesStore:        cesStore,
+		cesEnabled:      cesEnabled,
+	}
+
+	return r, nil
+}
+
+// syncCESsOnStartup updates the cache of CID usage in CES for all the
+// existing CESs.
+func (r *reconciler) calcDesiredStateOnStartup() error {
+	r.syncCESsOnStartup()
+	return r.syncPodsOnStartup()
+}
+
+func (r *reconciler) syncCESsOnStartup() {
+	if !r.cesEnabled {
+		return
+	}
+
+	for _, ces := range r.cesStore.List() {
+		r.cidUsageInCES.ProcessCESUpsert(ces.Name, ces.Endpoints)
+	}
+}
+
+// syncPodsOnStartup ensures that all pods have a CID for their labels, and that
+// all non-used CIDs are deleted. Non used CIDs are those that aren't in use by
+// any of the pods and also don't exist in CESs (if CES is enabled).
+func (r *reconciler) syncPodsOnStartup() error {
+	var lastError error
+
+	for _, pod := range r.podStore.List() {
+		if err := r.reconcilePod(podResourceKey(pod.Name, pod.Namespace)); err != nil {
+			lastError = err
+		}
+	}
+
+	return lastError
+}
+
+// reconcileCID ensures that the desired state for the CID is reached, by
+// comparing the CID in desired state cache and watcher's store and doing one of
+// the following:
+// 1. Nothing - If CID doesn't exist in both desired state cache and watcher's
+// store.
+// 2. Deletes CID - If CID only exists in the watcher's store, and it isn't used.
+// 3. Creates CID - If CID only exists in the desired state cache.
+// 4. Updates CID - If CIDs in the desired state cache and watcher's store are
+// not the same.
+func (r *reconciler) reconcileCID(cidResourceKey resource.Key) error {
+	cidName := cidResourceKey.Name
+	storeCID, existsInStore, err := r.cidStore.GetByKey(cidResourceKey)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	cidKey, existsInDesiredState := r.desiredCIDState.LookupByID(cidName)
+	if !existsInDesiredState && !existsInStore {
+		_ = r.makeIDAvailable(cidName)
```

Here we should be able to differentiate between the cases:

- `makeIDAvailable` failed because the identity was not previously reserved in the allocator. Nothing to do in that case.
- `makeIDAvailable` failed for unexpected reasons. Here we should propagate the error

---

I've added a log message to track this, I don't believe we need to propagate the error from reconcileCID.
If it would throw an error, we'll simply re-queue the cid reconciliation and it'll fail again. 

The potential failure reasons:
1) cidName is not an int (shouldn't happen)
2) cid is not in the range (shouldn't happen)
3) cid is there already so we shouldn't anyway propagate it (shouldn't happen)

---

Let's go with the warning, then! :+1: